### ISSUE TITLE
Add config file

### ISSUE THREAD

**Issue by Benaiah** [asciimoo/wuzz#38]
I think we're reaching the point where wuzz could use a config file. For syntax, I suggest [TOML](https://github.com/toml-lang/toml), which I've used before and is quite easy to integrate into Go.

There are a couple things that need to be decided on for the config file:

- Config sections and keys
- Available settings
- Default config location
- Command-line flag to specify a different config location

**Comment by Benaiah** [comment]
I believe `~/.config/wuzz/*` would be the proper location for the config file on Linux, but I'm not sure what the conventions are on other platforms. Should we make this platform-dependent, or should we go the typical `~/.wuzz/*` route, or even `~/.<wuzz config file>` (I think we should go for a directory, as there is potentially other stuff we'll want to store there).

**Comment by Eraden** [comment]
Yes, for most Linux `~/.config/APP/*` is current way to do this. In the past, home directory was polluted by countless rc files and subdirectories. It was terrible. Many application move to new convention like neovim, neomutt and even AwesomeWM.

**Comment by asciimoo** [comment]
I also prefer `~/.config/wuzz/`

**Comment by asciimoo** [comment]
Created a wikipage with the first draft of the config: https://github.com/asciimoo/wuzz/wiki/Config-file-draft . Feel free to edit/extend

**Comment by andybradshaw** [comment]
While `~/.config` is likely fine, respecting [`$XDG_CONFIG_HOME`](https://specifications.freedesktop.org/basedir-spec/basedir-spec-0.6.html) if it's defined on Linux is probably a good idea.

**Comment by georgijd** [comment]
The home directory seems the more natural choice to me as most people have means of dumping all their config files in there. As to file format, anything would do, really. TOML seems perfectly fine and we could support YAML as well, if there's a demand for that.

**Comment by Benaiah** [comment]
@asciimoo I've made an edit to my version of the wiki with a suggested modification to the config file format - see https://github.com/Benaiah/wuzz/wiki/Config-file-draft, moving the strings under "keys" to "keys.global". The commit message explains my thought process, but it's a pain to view wiki commits so I've copied it here:

---

## Change "keys" table to use "keys.global" instead of bare strings

The TOML library I've been using (https://github.com/BurntSush/toml) parses TOML into a Go struct, which is handy as it allows for a fully typed config values, so you can do something like `Config.DefaultTimeout` instead of something like `Config.Get("defaultTimeout")`. The reason for the change is because this requires that the TOML layout be expressible with Go's types - one of the major differences between Go's type system and TOML's spec is that Go maps cannot have disparate values.

Due to this, the previous keys layout would result in a type of `interface{}` for `Config.Keys`, eliminating any type-checking within it. The proposed structure would let us instead use the type `map[string] map[string]string`. It's a little confusing at first sight, but it just means you would access it using the following style:

```go
Config.Keys["global"]["C-n"]
Config.Keys["responseBody"]["C-s"]
```

This would remove the requirement to use either purely string-based configuration access or casting of `Config.Keys` to either `map[string] string` or `map[string] map[string]string` depending on whether you were accessing global or view-specific keys.

---

/end commit message

In other news, I have an initial config PR nearly ready to go, should show up in the next day or two. I'm going to leave out keyhandling initially so we can get new development using config, as it's more work to do the config stuff with every additional thing that needs to be retrofitted to use it.

**Comment by asciimoo** [comment]
> @asciimoo I've made an edit to my version of the wiki with a suggested modification to the config file format [..] moving the strings under "keys" to "keys.global"

Totally agree, it makes sense.

> In other news, I have an initial config PR nearly ready to go; it should show up in the next day or two.

Oh wow! I'm excited about it =]

> I'm going to leave out keybindings initially so we can get new development using the config system

Sure, it is a huge step and great addition without the keybindings too, and we can implement that later

**Comment by Benaiah** [comment]
> @Benaiah feel free to edit the wiki here, it is just a draft. Worst case, we'll revert the modifications.

Cool! I've pushed my wiki changes, and the config branch continues to go pretty well.

**Comment by Benaiah** [comment]
I've submitted my config branch as PR #51! I'd appreciate any help testing of some of the edge cases - particular areas of note include the command-line argument for specifying a config file (`-c <file>` or `--config <file>`), as it's parsed differently than the other arguments; invalid values for the various config settings; and different combinations of internal logic errors. Thanks everybody!

**Comment by Benaiah** [comment]
This has been closed by the merge of PR #51.
